### PR TITLE
Remove non-canonical usage of Deref

### PR DESCRIPTION
**Pull request by CodeSandwich**
Fixes https://github.com/input-output-hk/rust-cardano/issues/329

### CODE REVIEW

File: cardano/src/block/normal.rs [hunk 1]
```diff
@@ -104,8 +104,9 @@ impl cbor_event::de::Deserialize for DlgProof {
 
 #[derive(Debug, Clone)]
 pub struct TxPayload {
-    txaux: Vec<tx::TxAux>,
+    pub txaux: Vec<tx::TxAux>,
```
Comment (mzabaluev) [inline cardano/src/block/normal.rs]: I'd rather not expose internal members. In fact, I don't understand the purpose of this type which is just a wrapper over `Vec<tx::TxAux>`. Perhaps more principled accessor methods are needed on `Tx`, such as:

```rust
    fn contents(&self) -> &[tx::TxAux] {
        // ...
    }
```

File: cardano/src/block/normal.rs [hunk 2]
```diff
@@ -543,7 +537,8 @@ impl cbor_event::de::Deserialize for DecShare {
 // BTreeMap<StakeholderId, VssCertificate> see
 // https://github.com/input-output-hk/cardano-sl/blob/005076eb3434444a505c0fb150ea98e56e8bb3d9/core/src/Pos/Core/Ssc/VssCertificatesMap.hs#L36-L44
 #[derive(Debug, Clone)]
-pub struct VssCertificates(Vec<VssCertificate>);
+pub struct VssCertificates(pub Vec<VssCertificate>);
```
Comment (mzabaluev) [inline cardano/src/block/normal.rs]: Same thing here; why is it even a newtype?

File: cardano/src/block/verify.rs [hunk 3]
```diff
@@ -323,18 +323,18 @@ impl Verify for VssCertificates {
     fn verify(&self, protocol_magic: ProtocolMagic) -> Result<(), Error> {
         // check that there are no duplicate VSS keys
         let mut vss_keys = BTreeSet::new();
-        if !self.iter().all(|x| vss_keys.insert(x.vss_key.clone())) {
+        if !self.0.iter().all(|x| vss_keys.insert(x.vss_key.clone())) {
```
Comment (mzabaluev) [inline cardano/src/block/verify.rs]: I don't like all these `.0` accessors. If a wrapper type has a purpose (i.e. it's not just a meaningless newtype over something that everybody would actually want to use anyway), it should hide its contents and provide a named accessor method or methods to the inner value, `as`-style if accessing by reference, `into`-style if consumed to get the inner object.

File: cardano/src/block/verify.rs [hunk 4]
```diff
@@ -616,7 +616,7 @@ mod tests {
             if let Block::MainBlock(mblk) = &mut blk {
                 match &mut mblk.body.ssc {
                     normal::SscPayload::CommitmentsPayload(_, vss_certs) => {
-                        vss_certs[0].expiry_epoch = 123;
+                        vss_certs.0[0].expiry_epoch = 123;
```
Comment (mzabaluev) [inline cardano/src/block/verify.rs]: Ugh.

File: cardano/src/util/arbitrary.rs [hunk 5]
```diff
@@ -75,20 +71,20 @@ impl Arbitrary for Wrapper<(hdwallet::XPrv, address::ExtendedAddr)> {
...
             derivation_path: None,
             stake_distribution: address::StakeDistribution::BootstrapEraDistr,
-            network_magic: *network_magic,
+            network_magic: network_magic.into_inner(),
         };
         let addr_type = address::AddrType::ATPubKey;
         let addr = address::HashedSpendingData::new(
             addr_type,
-            &address::SpendingData::PubKeyASD(xprv.public()),
+            &address::SpendingData::PubKeyASD(xprv.clone().into_inner().public()),
```
Comment (mzabaluev) [inline cardano/src/util/arbitrary.rs]: It's more performant to have a by-ref accessor here and clone the inner object: `xprv.wotsit().clone()`

File: cardano/src/wallet/bip44.rs [hunk 6]
```diff
@@ -486,38 +463,36 @@ impl ChangeLevel<XPub> {
...
 }
-impl<T> Deref for ChangeLevel<T> {
-    type Target = T;
-    fn deref(&self) -> &T {
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct IndexLevel<T>(T);
+
+impl<T> IndexLevel<T> {
+    pub fn get_key(&self) -> &T {
```
Comment (mzabaluev) [inline cardano/src/wallet/bip44.rs]: Please be Rust style compliant and name the getter method just `key`.

File: cardano/src/wallet/rindex.rs [hunk 7]
```diff
@@ -382,16 +377,14 @@ impl RootKey {
         self.root_key
     }
 
+    pub fn get_xprv(&self) -> &XPrv {
```
Comment (mzabaluev) [inline cardano/src/wallet/rindex.rs]: Same here, `xprv`.

File: chain-impl-mockchain/src/multiverse.rs [hunk 8]
```diff
@@ -62,12 +62,9 @@ impl GCRoot {
 
         GCRoot { hash, roots }
     }
-}
 
-impl std::ops::Deref for GCRoot {
-    type Target = BlockId;
-    fn deref(&self) -> &Self::Target {
-        &self.hash
+    pub fn get_hash(&self) -> BlockId {
```
Comment (mzabaluev) [inline chain-impl-mockchain/src/multiverse.rs]: Just `hash`, or maybe `id`?

File: exe-common/src/config.rs [hunk 9]
```diff
@@ -234,17 +223,17 @@ pub mod net {
...
+                .iter()
                 .filter_map(|np| np.peer().get_native())
                 .collect()
         }
-    }
-    impl Deref for Peers {
-        type Target = Vec<NamedPeer>;
-        fn deref(&self) -> &Self::Target {
+
+        pub fn get_peers(&self) -> &[NamedPeer] {
```
Comment (mzabaluev) [inline exe-common/src/config.rs]: `peers`

File: exe-common/src/network/native.rs [hunk 10]
```diff
@@ -96,34 +93,30 @@ impl Api for PeerPool {
...
-impl Deref for Connection {
-    type Target = OpenPeer;
-    fn deref(&self) -> &Self::Target {
-        &self.1
-    }
-}
-impl DerefMut for Connection {
-    fn deref_mut(&mut self) -> &mut Self::Target {
+
+    pub fn get_peer_mut(&mut self) -> &mut OpenPeer {
```
Comment (mzabaluev) [inline exe-common/src/network/native.rs]: `peer_mut`

**Comment by vincenthz** [comment]
closing since it doesn't apply anymore, and it would make sense to split this into smaller / more digestable piece. At the moment, I think we should all keep this in mind in the new code and prevent `Deref` to be added needlessly. Changing the actual interface in `cardano` is somewhat painful (multi repository to check/adjust), so it might be better to do this when we have time for janitorial work

**Pull request closed without merging**